server/src/test/java/com/google/fhir/gateway/AuditEventHelperTest.java

```diff
@@ -0,0 +1,736 @@
+    String fullRequestUrl =
+        String.format(
+            "%s/Observation?_id=test-observation-id-1&_id=test-observation-id-2",
+            FHIR_SERVER_BASE_URL);
+
+    when(requestDetailsReader.getFhirServerBase()).thenReturn(FHIR_SERVER_BASE_URL);
+    when(requestDetailsReader.getRequestType()).thenReturn(RequestTypeEnum.GET);
+    when(requestDetailsReader.getCompleteUrl()).thenReturn(fullRequestUrl);
+    when(requestDetailsReader.getRestOperationType()).thenReturn(RestOperationTypeEnum.SEARCH_TYPE);
+    when(requestDetailsReader.getCompleteUrl()).thenReturn(fullRequestUrl);
+    when(requestDetailsReader.getRequestPath()).thenReturn("Observation");
+
+    when(requestDetailsReader.getParameters())
+        .thenReturn(Map.of("_id", new String[] {"test-observation-id-1", "test-observation-id-2"}));
+
+    Bundle responseBundle = new Bundle();
+    responseBundle.setType(Bundle.BundleType.TRANSACTIONRESPONSE);
+    responseBundle.setId("response-bundle-1");
+
+    Bundle.BundleEntryComponent postResponseEntry = new Bundle.BundleEntryComponent();
+    Bundle.BundleEntryResponseComponent postResponse = new Bundle.BundleEntryResponseComponent();
+    postResponse.setLocation("Observation/test-observation-id-1");
+    postResponseEntry.setResponse(postResponse);
+    responseBundle.addEntry(postResponseEntry);
+
+    Bundle.BundleEntryComponent putResponseEntry = new Bundle.BundleEntryComponent();
+    Bundle.BundleEntryResponseComponent putResponse = new Bundle.BundleEntryResponseComponent();
+    putResponse.setLocation("Observation/test-observation-id-2");
+    putResponseEntry.setResponse(putResponse);
+    responseBundle.addEntry(putResponseEntry);
+
+    AuditEventHelper auditEventHelper =
+        createTestInstance(null, responseBundle, RestOperationTypeEnum.SEARCH_TYPE);
+    auditEventHelper.processAuditEvents();
+
+    ArgumentCaptor<IBaseResource> payloadResourceCaptor =
+        ArgumentCaptor.forClass(IBaseResource.class);
+
+    verify(fhirClientMock, times(2)).postResource(payloadResourceCaptor.capture());
+
+    List<IBaseResource> resources = payloadResourceCaptor.getAllValues();
+
+    assertThat(resources.size(), is(2));
+    assertThat(resources.get(0) instanceof AuditEvent, is(true));
+    assertThat(resources.get(1) instanceof AuditEvent, is(true));
+
+    AuditEvent auditEvent;
+
+    // Observation 1
+    auditEvent = (AuditEvent) resources.get(0);
+    assertCommonAuditEventFields(auditEvent, false);
+    assertThat(auditEvent.getAction().toCode(), equalTo("E"));
+    assertThat(auditEvent.getSubtype().get(0).getCode(), equalTo("search-type"));
+    assertThat(
+        getAuditEventDomainResourceReference(auditEvent.getEntity()),
+        equalTo("Observation/test-observation-id-1"));
+    assertThat(
+        getEntityQueryBySystemCode(
+            auditEvent.getEntity(), "http://terminology.hl7.org/CodeSystem/object-role", "24"),
+        equalTo(fullRequestUrl));
```

Comment:
Shouldn't this value be base64 encoded? I mean, IIUC the `entity.query` field should be based64 encoded, so I wonder how does this check pass?

Reply:
The HAPI FHIR structure returns a byte array and we convert it to a string when we return from the method:
```java
return new String(component.getQuery(), StandardCharsets.UTF_8);
```